firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java

```diff
@@ -357,6 +458,30 @@ private Query whereHelper(@NonNull FieldPath fieldPath, Operator op, Object valu
                 + Util.typeName(value));
       }
     } else {
+      if (op == Operator.IN || op == Operator.ARRAY_CONTAINS_ANY) {
+        if (!(value instanceof List) || ((List) value).size() == 0) {
+          throw new IllegalArgumentException(
+              "Invalid Query. A non-empty array is required for '" + op.toString() + "' filters.");
+        }
+        if (((List) value).size() > 10) {
+          throw new IllegalArgumentException(
+              "Invalid Query. '"
+                  + op.toString()
+                  + "' filters support a maximum of 10 elements in the value array.");
+        }
+        if (((List) value).contains(null)) {
+          throw new IllegalArgumentException(
+                  "Invalid Query. '"
+                          + op.toString()
+                          + "' filters cannot contain 'null' in the value array.");
+        }
+        if (((List) value).contains(Double.NaN) || ((List) value).contains(Float.NaN)) {
```

Comment:
Does this really work? I thought that contains uses `equals` and `NaN` never equals itself?

Reply:
This seems to work based on the unit tests: https://github.com/firebase/firebase-android-sdk/pull/519/files#diff-37f4386134136e6f82e850b93ff16f8fR659

I verified that removing either `Double.NaN` or `Float.NaN` from the `if` will cause the corresponding test to fail.

Reply:
That seems surprising, does this work with other kinds of NaN?

Here are some sample values:

```

  private static final Double NEGATIVE_NAN = Double.longBitsToDouble(0xFFF1100000000000L);
  private static final Double SIGNALLING_NAN = Double.longBitsToDouble(0x7FF1000000000000L);
  private static final Double NEGATIVE_SIGNALLING_NAN =
      Double.longBitsToDouble(0xFFF1000000000000L);
  private static final Double OTHER_NAN = Double.longBitsToDouble(0x7FF1000000000001L);

  private static final Double NEGATIVE_OTHER_NAN = Double.longBitsToDouble(0xFFF1000000000001L);
```

Maybe this is overkill tho? I don't know the JVM spec exactly for NaNs and if they are normalized or not.